agent/main.go

```diff
@@ -34,6 +36,15 @@ import (
 
 const defaultPort = 10789
 
+// A fake publisher that we use as a bandaid to prevent nil pointer exceptions when the
+// runc shim tries to publish events.
+// TODO replace this with a publisher that forwards events over vsock to the outer shim
```

Comment:
Just wanted to add some context here: [originally](https://github.com/containerd/containerd/blob/26ab393e7ddf23d245244043ae4629a2a57dc453/runtime/v2/runc/service.go#L663) (before Jan 25, 2019) there was just one event published by `runc` - `TaskExit` event, so we decided to go with [pull model](https://github.com/firecracker-microvm/firecracker-containerd/blob/7f37fafbadb4789decd6bf34ae74283e189f0d62/runtime/service.go#L226) as it seemed like an overkill to implement a fully-featured publisher for an exit event.

There is a [commit](https://github.com/containerd/containerd/commit/85aa8ad3614ed183f6db8564140cdc876d2facee#diff-6238adea42c707e286831f2a8566d37d) which adds a lot more events to `runc`, that was added about one month ago. I think this is when we started getting `nil` errors.

Reply:
Ah interesting, good to know!